rustls-post-quantum/src/lib.rs

```diff
@@ -25,10 +25,184 @@ pub fn provider() -> CryptoProvider {
     #[cfg(feature = "aws-lc-rs-unstable")]
     {
         provider.signature_verification_algorithms = SUPPORTED_SIG_ALGS;
+        provider.key_provider = &key_provider::PqAwsLcRs;
     }
     provider
 }
 
+#[cfg(feature = "aws-lc-rs-unstable")]
+mod key_provider {
+    #[cfg(feature = "aws-lc-rs-unstable")]
+    use std::fmt::{self, Debug, Formatter};
+    #[cfg(feature = "aws-lc-rs-unstable")]
+    use std::sync::Arc;
+
+    use aws_lc_rs::signature::KeyPair;
+    #[cfg(feature = "aws-lc-rs-unstable")]
+    use aws_lc_rs::unstable::signature::{
+        ML_DSA_44_SIGNING, ML_DSA_65_SIGNING, ML_DSA_87_SIGNING, PqdsaKeyPair,
+        PqdsaSigningAlgorithm,
+    };
+    use rustls::crypto::KeyProvider;
+    use rustls::crypto::aws_lc_rs::sign;
+    use rustls::pki_types::{AlgorithmIdentifier, PrivateKeyDer, SubjectPublicKeyInfoDer, alg_id};
+    use rustls::sign::{Signer, SigningKey};
+    use rustls::{Error, SignatureAlgorithm, SignatureScheme};
+
+    #[derive(Debug)]
+    pub(super) struct PqAwsLcRs;
+
+    impl KeyProvider for PqAwsLcRs {
+        fn load_private_key(
+            &self,
+            key_der: PrivateKeyDer<'static>,
+        ) -> Result<Arc<dyn SigningKey>, Error> {
+            const ML_DSA_SIGNING_ALGS: &[&PqdsaSigningAlgorithm] =
+                &[&ML_DSA_44_SIGNING, &ML_DSA_65_SIGNING, &ML_DSA_87_SIGNING];
+
+            for &alg in ML_DSA_SIGNING_ALGS {
+                let key_pair = match &key_der {
+                    PrivateKeyDer::Pkcs8(pkcs8) => {
+                        match PqdsaKeyPair::from_pkcs8(alg, pkcs8.secret_pkcs8_der()) {
+                            Ok(key_pair) => key_pair,
+                            Err(_) => continue,
+                        }
+                    }
+                    // To do: support `PqdsaKeyPair::from_raw_private_key()`?
+                    _ => {
+                        return Err(Error::General(
+                            "ML-DSA keys must be provided in PKCS#8 format".into(),
+                        ));
+                    }
+                };
+
+                return Ok(Arc::new(PqdsaSigningKey {
+                    kind: if alg == &ML_DSA_44_SIGNING {
+                        PqdsaKeyKind::MlDsa44
+                    } else if alg == &ML_DSA_65_SIGNING {
+                        PqdsaKeyKind::MlDsa65
+                    } else if alg == &ML_DSA_87_SIGNING {
+                        PqdsaKeyKind::MlDsa87
+                    } else {
+                        unreachable!("unexpected ML-DSA signing algorithm")
+                    },
+                    inner: Arc::new(key_pair),
+                }));
+            }
+
+            if let Ok(key) = sign::any_supported_type(&key_der) {
+                return Ok(key);
+            }
+
+            Err(Error::General(
+                "failed to parse private key as ML-DSA, RSA, ECDSA, or EdDSA".into(),
+            ))
```

Comment:
nit: would a `match` be more idiomatic here?

```rust
            match sign::any_supported_type(&key_der) {
                Ok(key) => Ok(key),
                Err(_) => Err(Error::General(
                    "failed to parse private key as ML-DSA, RSA, ECDSA, or EdDSA".into(),
                )),
            }
```

Reply:
Yup.